keystore/keystore.go

```diff
@@ -0,0 +1,123 @@
+package keystore
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	ci "gx/ipfs/QmfWDLQjGjVe4fr5CoztYW2DYYjRysMJrFe1RCsXLPTf46/go-libp2p-crypto"
+)
+
+type Keystore interface {
+	Put(string, ci.PrivKey) error
+	Get(string) (ci.PrivKey, error)
+	Delete(string) error
+	List() ([]string, error)
+}
+
+var ErrNoSuchKey = fmt.Errorf("no key by the given name was found")
+var ErrKeyExists = fmt.Errorf("key by that name already exists, refusing to overwrite")
+
+type FSKeystore struct {
+	dir string
+}
+
+func validateName(name string) error {
+	if name == "" {
+		return fmt.Errorf("key names must be at least one character")
+	}
+
+	if strings.Contains(name, "/") {
+		return fmt.Errorf("key names may not contain slashes")
+	}
+
+	if strings.HasPrefix(name, ".") {
+		return fmt.Errorf("key names may not begin with a period")
+	}
+
+	return nil
+}
+
+func NewFSKeystore(dir string) (*FSKeystore, error) {
+	_, err := os.Stat(dir)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		if err := os.Mkdir(dir, 0700); err != nil {
+			return nil, err
+		}
+	}
+
+	return &FSKeystore{dir}, nil
+}
+
+func (ks *FSKeystore) Put(name string, k ci.PrivKey) error {
+	if err := validateName(name); err != nil {
+		return err
+	}
+
+	b, err := k.Bytes()
+	if err != nil {
+		return err
+	}
+
+	kp := filepath.Join(ks.dir, name)
+
+	_, err = os.Stat(kp)
+	if err == nil {
```

Comment:
To be safe I would use `os.IsExist` and change it to `ErrKeyExists` then, else passthrough the error.